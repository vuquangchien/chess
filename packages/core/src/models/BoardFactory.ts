import {Board} from './Board';
import {FenRecord} from './fen/FenRecord';

export class BoardFactory {
  public static createEmptyBoard(): Board {
    return new Board();
  }
  public static createFromFen(fen: string): Board {
    const board = new Board();
    const fenRecord = new FenRecord(fen);
    const cells = fenRecord.piecePlacement.getCells();
    cells.forEach(cell => {
      board.getByCellName(cell.name).setPiece(cell.piece);
    });
    board.setCurrentPlayer(fenRecord.activeColor);
    return board;
  }
  public static createBoardWithPieces(): Board {
    return BoardFactory.createFromFen(
      'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w - - 0 1'
    );
  }
}
