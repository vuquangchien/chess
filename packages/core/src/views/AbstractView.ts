import {IView} from '../framework/IView';
import {IEngine} from './IEngine';

export abstract class AbstractView<T> implements IView<T> {
  engine!: IEngine;
  children: AbstractView<unknown>[] = [];
  controller: T;

  public constructor(controller: T) {
    this.children = [];
    this.controller = controller;
    this.init();
  }
  init(): void {}

  setEngine(engine: IEngine): void {
    this.engine = engine;
    this.children.forEach(child => child.setEngine(engine));
  }
  addChild(child: AbstractView<unknown>): void {
    this.children.push(child);
  }

  addChildren(children: AbstractView<unknown>[]): void {
    this.children.push(...children);
  }

  renderChildren(): string {
    let result = '';
    for (const child of this.children) {
      result += child.render();
    }
    return result;
  }

  abstract render(): string;
}
