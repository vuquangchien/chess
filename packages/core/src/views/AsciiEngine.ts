import {IEngine} from './IEngine';

export class AsciiEngine implements IEngine {
  renderBox(options: {background: string; text: string}): string {
    const {background, text} = options;
    return (background === '#000' ? '■' : '□') + text;
  }
}
