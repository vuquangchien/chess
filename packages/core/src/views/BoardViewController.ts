import {Board} from '../models/Board';

export class BoardViewController {
  private board: Board;

  constructor(board: Board) {
    this.board = board;
  }

  public getBoard(): Board {
    return this.board;
  }
}
