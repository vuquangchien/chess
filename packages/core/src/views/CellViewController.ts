import {IViewController} from '../framework/IViewController';
import {Cell} from '../models/Cell';

export class CellViewController implements IViewController {
  public cell: Cell;

  constructor(cell: Cell) {
    this.cell = cell;
  }
}
