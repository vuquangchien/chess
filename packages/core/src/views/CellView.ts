import {CellViewController} from './CellViewController';
import {IEngine} from './IEngine';
import {EngineFactory} from './EngineFactory';
import {AbstractView} from './AbstractView';

const Icons = {
  rook: {
    white: '♖',
    black: '♜',
  },
  knight: {
    white: '♘',
    black: '♞',
  },
  bishop: {
    white: '♗',
    black: '♝',
  },
  queen: {
    white: '♕',
    black: '♛',
  },
  king: {
    white: '♔',
    black: '♚',
  },
  pawn: {
    white: '♙',
    black: '♟',
  },
};
export class CellView extends AbstractView<CellViewController> {
  controller: CellViewController;
  engine: IEngine;

  constructor(controller: CellViewController) {
    super(controller);
    this.controller = controller;
    this.engine = EngineFactory.createEngine('ascii');
  }

  render(): string {
    const cell = this.controller.cell;
    let result = '';
    const piece = cell.getPiece();
    const background = cell.color === 'black' ? '#D18B47' : '#FFCE9E';
    const icon = piece ? Icons[piece.getType()][piece.getColor()] : ' ';
    result = this.engine.renderBox({
      column: cell.colIndex,
      row: cell.rowIndex,
      background,
      text: icon,
    });

    // if (piece === null) {
    //   if (!cell.isHighlighted()) {
    //     result = this.engine.renderBox({
    //       column: cell.file,
    //       row: cell.rank,
    //       background,
    //       text: ' ',
    //     });
    //   } else {
    //     result = '× ';
    //   }
    // } else {
    //   const icon = Icons[piece.getType()][piece.getColor()];
    //   if (cell.isHighlighted()) {
    //     result = '×' + icon;
    //   } else result = icon + ' ';
    // }
    return result;
  }
}
