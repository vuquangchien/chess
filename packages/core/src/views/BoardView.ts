import {CellView} from './CellView';
import {CellViewController} from './CellViewController';

import {BoardViewController} from './BoardViewController';
import {RowEl, TextEl} from './RowEl';
import {AbstractView} from './AbstractView';

export class BoardView extends AbstractView<BoardViewController> {
  public init(): void {
    this.addChild(new RowEl());
    const headerRow = new RowEl();
    headerRow.addChildren(
      '   a b c d e f g h'
        .split(' ')
        .map(el => new TextEl({text: el + '', row: 0, column: 0}))
    );
    this.addChild(headerRow);
    const board = this.controller.getBoard();
    const iterator = board.getRankIterator();
    while (iterator.hasNext()) {
      const row = new RowEl();
      row.addChild(new TextEl({text: `${7 - iterator.currentIndex()} `}));
      const rank = iterator.next();
      const it = rank.getIterator();
      while (it.hasNext()) {
        const cell = it.next();
        const vc = new CellViewController(cell);
        const cellView = new CellView(vc);
        row.addChild(cellView);
      }
      this.addChild(row);
    }
  }

  public render(): string {
    return this.renderChildren();
  }
}
