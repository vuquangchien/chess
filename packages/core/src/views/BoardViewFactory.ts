import {Board} from '../models/Board';
import {BoardViewController} from './BoardViewController';
import {BoardView} from './BoardView';

export class BoardViewFactory {
  public static fromBoard(board: Board): BoardView {
    return new BoardView(new BoardViewController(board));
  }
}
