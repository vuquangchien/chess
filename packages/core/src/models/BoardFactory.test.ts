import {BoardFactory} from './BoardFactory';
import {BoardViewFactory} from '../views/BoardViewFactory';
// @ts-ignore
import {expect} from '@jest/globals';

describe('BoardFactory', () => {
  it('should create empty board', () => {
    const board = BoardFactory.createEmptyBoard();
    const render = BoardViewFactory.fromBoard(board).render();
    expect(render).toMatchSnapshot();
  });
  it('should create board from fen', () => {
    const board = BoardFactory.createFromFen(
      'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w - - 0 1'
    );
    const render = BoardViewFactory.fromBoard(board).render();
    expect(render).toMatchSnapshot();
  });
  it('should create board with pieces', () => {
    const board = BoardFactory.createBoardWithPieces();
    const render = BoardViewFactory.fromBoard(board).render();
    expect(render).toMatchSnapshot();
  });
});
