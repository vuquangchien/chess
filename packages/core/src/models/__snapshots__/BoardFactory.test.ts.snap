// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardFactory should create board from fen 1`] = `
"
  a b c d e f g h
8 ♖ ♘ ♗ ♕ ♔ ♗ ♘ ♖ 
7 ♙ ♙ ♙ ♙ ♙ ♙ ♙ ♙ 
6 □ ■ □ ■ □ ■ □ ■ 
5 ■ □ ■ □ ■ □ ■ □ 
4 □ ■ □ ■ □ ■ □ ■ 
3 ■ □ ■ □ ■ □ ■ □ 
2 ♟ ♟ ♟ ♟ ♟ ♟ ♟ ♟ 
1 ♜ ♞ ♝ ♛ ♚ ♝ ♞ ♜ 
"
`;

exports[`BoardFactory should create empty board 1`] = `
"
  a b c d e f g h
8 □ ■ □ ■ □ ■ □ ■ 
7 ■ □ ■ □ ■ □ ■ □ 
6 □ ■ □ ■ □ ■ □ ■ 
5 ■ □ ■ □ ■ □ ■ □ 
4 □ ■ □ ■ □ ■ □ ■ 
3 ■ □ ■ □ ■ □ ■ □ 
2 □ ■ □ ■ □ ■ □ ■ 
1 ■ □ ■ □ ■ □ ■ □ 
"
`;
